kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -107,31 +118,42 @@ private class SemaphoreImpl(
     private val enqIdx = atomic(0L)
     private val deqIdx = atomic(0L)
 
-    override fun tryAcquire(): Boolean {
-        _availablePermits.loop { p ->
-            if (p <= 0) return false
-            if (_availablePermits.compareAndSet(p, p - 1)) return true
+    /**
+     * The remaining permits from release operations, which could not be spent, because the next slot was not defined
+     */
+    internal val accumulator = atomic(0)
+
+    override fun tryAcquire(permits: Int): Boolean {
+        require(permits > 0) { "The number of acquired permits must be greater than 0" }
+        permitsBalance.loop { p ->
+            if (p < permits) return false
+            if (permitsBalance.compareAndSet(p, p - permits)) return true
         }
     }
 
-    override suspend fun acquire() {
-        val p = _availablePermits.getAndDecrement()
+    override suspend fun acquire(permits: Int) {
+        require(permits > 0) { "The number of acquired permits must be greater than 0" }
+        val p = _availablePermits.getAndAdd(-permits)
         if (p > 0) return // permit acquired
         addToQueueAndSuspend()
     }
 
-    override fun release() {
-        val p = incPermits()
+    override fun release(permits: Int) {
+        require(permits > 0) { "The number of released permits must be greater than 0" }
+        val p = incPermits(permits)
         if (p >= 0) return // no waiters
-        resumeNextFromQueue()
+        repeat(permits) {
+            resumeNextFromQueue()
+        }
     }
 
-    fun incPermits() = _availablePermits.getAndUpdate { cur ->
-        check(cur < permits) { "The number of released permits cannot be greater than $permits" }
-        cur + 1
+    internal fun incPermits(delta: Int = 1) = permitsBalance.getAndUpdate { cur ->
+        assert { delta >= 1 }
+        check(cur + delta <= permits) { "The number of released permits cannot be greater than $permits" }
+        cur + delta
     }
 
-    private suspend fun addToQueueAndSuspend() = suspendAtomicCancellableCoroutine<Unit> sc@ { cont ->
+    private suspend fun tryToAddToQueue(permits: Int) = suspendAtomicCancellableCoroutine<Unit> sc@{ cont ->
```

Comment:
I don't see how `permits` is used by this method. I'm very surprised that tests pass. Seems like some tests are missing.

Reply:
The previous commit didn't contain significant changes, sorry. Please, recheck.